src/setup_payload/python/SetupPayload.py

```diff
@@ -0,0 +1,220 @@
+        self.long_discriminator = discriminator
+        self.short_discriminator = discriminator >> 8
+        self.pincode = pincode
+        self.discovery = rendezvous
+        self.flow = flow
+        self.vid = vid
+        self.pid = pid
+
+    def p_print(self):
+        print('{:<{}} :{}'.format('Flow', 24, self.flow))
+        print('{:<{}} :{}'.format('Pincode', 24, self.pincode))
+        print('{:<{}} :{}'.format('Short Discriminator', 24, self.short_discriminator))
+        if self.long_discriminator:
+            print('{:<{}} :{}'.format('Long Discriminator', 24, self.long_discriminator))
+        if self.discovery:
+            print('{:<{}} :{}'.format('Discovery Capabilities', 24, self.discovery))
+        if self.vid is not None and self.pid is not None:
+            print('{:<{}} :{:<{}} (0x{:04x})'.format('Vendor Id', 24, self.vid, 6, self.vid))
+            print('{:<{}} :{:<{}} (0x{:04x})'.format('Product Id', 24, self.pid, 6, self.pid))
+
+    def qrcode_dict(self):
+        return {
+            'version': 0,
+            'vid': self.vid,
+            'pid': self.pid,
+            'flow': int(self.flow),
+            'discovery': self.discovery,
+            'discriminator': self.long_discriminator,
+            'pincode': self.pincode,
+            'padding': 0,
+        }
+
+    def manualcode_dict(self):
+        return {
+            'version': 0,
+            'vid_pid_present': 0 if self.flow == CommissioningFlow.Standard else 1,
+            'discriminator': self.short_discriminator,
+            'pincode_lsb': self.pincode & 0x3FFF,    # 14 ls-bits
+            'pincode_msb': self.pincode >> 14,       # 13 ms-bits
+            'vid': 0 if self.flow == CommissioningFlow.Standard else self.vid,
+            'pid': 0 if self.flow == CommissioningFlow.Standard else self.pid,
+            'padding': 0,
+        }
+
+    def generate_qrcode(self):
+        data = qrcode_format.build(self.qrcode_dict())
+        b38_encoded = Base38.encode(data[::-1])  # reversing
+        return 'MT:{}'.format(b38_encoded)
+
+    def generate_manualcode(self):
+        CHUNK1_START = 0
+        CHUNK1_LEN = 4
+        CHUNK2_START = CHUNK1_START + CHUNK1_LEN
+        CHUNK2_LEN = 16
+        CHUNK3_START = CHUNK2_START + CHUNK2_LEN
+        CHUNK3_LEN = 13
+
+        bytes = manualcode_format.build(self.manualcode_dict())
+        bits = bitarray()
+        bits.frombytes(bytes)
+
+        chunk1 = str(int(bits[CHUNK1_START:CHUNK1_START + CHUNK1_LEN].to01(), 2)).zfill(1)
+        chunk2 = str(int(bits[CHUNK2_START:CHUNK2_START + CHUNK2_LEN].to01(), 2)).zfill(5)
+        chunk3 = str(int(bits[CHUNK3_START:CHUNK3_START + CHUNK3_LEN].to01(), 2)).zfill(4)
+        chunk4 = str(self.vid).zfill(5) if self.flow != CommissioningFlow.Standard else ''
+        chunk5 = str(self.pid).zfill(5) if self.flow != CommissioningFlow.Standard else ''
```

Comment:
Why `str` rather than `bytes` here?

Reply:
At the end we have to convert the numbers to string and to me this seems easier as we can pad the zeros using zfill().